Arrowgene.Ddon.GameServer/Characters/HubManager.cs

```diff
@@ -0,0 +1,164 @@
+using Arrowgene.Ddon.Server;
+using Arrowgene.Ddon.Shared.Entity.PacketStructure;
+using Arrowgene.Logging;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Arrowgene.Ddon.GameServer.Characters
+{
+    public class HubManager
+    {
+        private static readonly ServerLogger Logger = LogProvider.Logger<ServerLogger>(typeof(HubManager));
+
+        public HubManager(DdonGameServer server)
+        {
+            Server = server;
+            HubMembers = new Dictionary<uint, HashSet<GameClient>>();
+            foreach (var stageId in StageManager.HubStageIds)
+            {
+                HubMembers[stageId] = new HashSet<GameClient>();
+            }
+        }
+
+        private readonly DdonGameServer Server;
+        private readonly Dictionary<uint, HashSet<GameClient>> HubMembers;
+
+        //The server maintains an authoritative list of who's in each hub stage.
+        //Hub stages are defined in StageManager.HubStageIds.
+        //The client has a weird way of keeping contexts, which I attempt to replicate here.
+        //As long as the client does not see another client leave a hub, it preserves that other client's context
+        //in memory and does not need to get it again from the server to see that client's character.
+
+        public void MoveStageUpdateLastSeen(GameClient client, uint sourceStageId, uint targetStageId)
+        {
+            //Fallback to classic method.
+            if (!Server.Setting.GameLogicSetting.FancyLobbyContextHandling)
+            {
+                ClassicLobbyHandling(client, sourceStageId);
+                return;
+            }
+
+            //Transitions that do not involve a hub stage don't concern us.
+            if (!HubMembers.ContainsKey(sourceStageId) && !HubMembers.ContainsKey(targetStageId)) return;
```

Comment:
Can we put these 1 line if statements into a block. It is hard to notice sometimes and also prone to error.

```c#
if (!HubMembers.ContainsKey(sourceStageId) && !HubMembers.ContainsKey(targetStageId))
{
    return;
}
```